Sources/Support/StoreMessagesHelper.swift

```diff
@@ -0,0 +1,187 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  StoreMessagesHelper.swift
+//
+//  Created by Antonio Rico Diez on 27/9/23.
+
+import StoreKit
+
+class StoreMessagesHelper {
+
+    private let systemInfo: SystemInfo
+    private let showStoreMessagesAutomatically: Bool
+    private let storeMessagesProvider: StoreMessagesProvider
+
+    // WIP: Need to see how this handles concurrency
+    private var deferredMessages: [StoreMessage] = []
+
+    init(systemInfo: SystemInfo,
+         showStoreMessagesAutomatically: Bool) {
+        self.systemInfo = systemInfo
+        self.showStoreMessagesAutomatically = showStoreMessagesAutomatically
+        self.storeMessagesProvider = StoreMessagesProviderWrapper()
+    }
+
+    @available(iOS 16.0, *)
+    init(systemInfo: SystemInfo,
+         showStoreMessagesAutomatically: Bool,
+         storeMessagesProvider: StoreMessagesProvider = StoreMessagesProviderWrapper()) {
+        self.systemInfo = systemInfo
+        self.showStoreMessagesAutomatically = showStoreMessagesAutomatically
+        self.storeMessagesProvider = storeMessagesProvider
+    }
+
+    #if os(iOS)
+
+    @available(iOS 16.4, *)
+    @available(macOS, unavailable)
+    @available(watchOS, unavailable)
+    @available(tvOS, unavailable)
+    func deferMessagesIfNeeded() async throws {
+        guard !self.showStoreMessagesAutomatically else {
+            return
+        }
+        Task(priority: .background) {
```

Comment:
I moved this inside a background task. Otherwise this will keep listening for messages forever, getting stuck. This did cause some issues in the tests though, since we have to wait until the messages are cached before showing them.

Reply:
We may want to perform the append in the main thread just to be safe since we're accessing it from the main actor in `showStoreMessages`. Something like:

```
await MainActor.run {
    self.deferredMessages.append(message)
}
```

Reply:
Also, I think there may be a retain cycle here, we can break it by making self weak inside the task:

`Task(priority: .background) { [weak self]`

and then calling it optionally:

`self?.deferredMessages.append(message)`